ledger-ondisk/src/LedgerOnDisk/Diff.hs

```diff
@@ -38,29 +38,72 @@ data Diff v where
 deriving stock instance Eq v => Eq (Diff v)
 deriving stock instance Show v => Show (Diff v)
 
+{-  (<>) on Diff is not commutative, so I think we should change the order of arguments here for 2 reasons
+ 1) I like to think of (d1 <> d2) as composition of changes. So in accordance with Haskell's composition
+    we should think of addpying the d2 first.  E.g. in Haskell (f . g) x == f (g x). So apply (x <> y) x == apply x (apply y x)
+ 2) When we lift (Diff v) to Map k (Diff v), we will want to use Map.insertWith (<>) k v x, and Map.unionWith (<>) m1 m2
+    and these require to us apply the change on the right first.
+So it is best to change the order.
+
 instance Semigroup (Diff v) where
   x <> DNoChange = x
   DMappend x <> DMappend y = DMappend $ x <> y
   DChangeTo x <> DMappend y = DChangeTo $ x <> y
   DRemove <> DMappend y = DChangeTo y
   _ <> y = y
 
+Another thing, is that DRemove (applied first) cancels a subsequent (DMappend v) so the 4th line
+when appying things in the eft first should be
+   DRemove <> DMappend y = DRemove.
+We can also make the function clearer by making it more verbose, so it is easy to see all the cases are covered.
+So I suggest this.
+-}
+
+-- (<>) is composition of changes. So apply (x <> y) x == apply x (apply y x)
+-- Note we apply the change on the right first.
+instance Semigroup v => Semigroup (Diff v) where
+  DNoChange <> x          = x
+  DRemove <> x            = DRemove
+  DChangeTo v <> x        = DChangeTo v
+  DMappend v <> DNoChange = DMappend v
+  DMappend v <> DRemove   = DRemove
+  DMappend v <> DChangeTo u = DChangeTo (v <> u)
+  DMappend v <> DMappend u = DMappend (v <> u)  
+
 instance Monoid (Diff v) where
   mempty = DNoChange
 
+
+-- Not sure this is a very useful function, it is just the 'fmap' method of functor.
```

Comment:
Agree it's not so useful. But it's the best version of fmap available , due to the DMappend constructor.
Undocumented precondition is that f is a Semigroup homomorphism.

Reply:
I want above all  that he law    (apply (d1 <> d2) m) == apply d1 (apply d2 m)      holds.   And I want there to be nice properties for fmap as well.   If fmap is the function part of the Functor, then the identity and composition laws hold. But they seem to be at odds with the apply laws in some strange way.  So I ask, do we ever use this function?   If not lets remove it. Its type suggests it s a functor, but it doesn't really behave as one.  If we do use it, lets be sure we are using it the correctly.

This reaches back to the comment you made about the use of functions in my Message type, and my reply that we could make first order representaitons of functions.  So lets see how that works out. Suppose we had the choice of using either

data Diff v =
  DMappend v | DFun (v -> v)

The we could represent (DMappend 3) as (DFun (3 <>))  and we would get the same answer.   We could also represent DNoChange  with (DFun id),  So this looks promissing, But we can't make a functor instance of a type with DFun here as well. given

data Diff v = DChangeTo v | DFun (v -> v)

fmap f (DChangeTo v) = DChangeTo (f v)
fmap f (DFun g) = DFun( f . g . h)
   where h is the inverse of f,  but how do we get that?

So a Functor instance seems impossible.  One last thought. Suppose f and h form an isomorphism, then we can write something approaching a Functor instance

isomap :: (Iso a b) -> Diff a -> Diff b
isomap (Iso f h)  DChangeTo (f v)
isomap (Iso f h) (DFun g) = DFun( f . g . h)

Reply:
I'm not attached to this function at all, but I do think it's fine if `f` is a semigroup homomorphism (i.e. `f (x <> y) == f x <> f y`). I think your `isomap` is also fine.